src/hbbft_acs.erl

```diff
@@ -161,24 +161,51 @@ handle_msg(Data = #acs_data{n=N, f=F}, J, {{bba, I}, BBAMsg}) ->
             {store_bba_state(Data, I, NewBBA), defer}
     end.
 
-check_completion(Data = #acs_data{n=N}, ToSend) ->
+check_completion(Data, ToSend) ->
     %% Figure4, Bullet4
-    %% once all instances of BA have completed, let C⊂[1..N] be the indexes of each BA that delivered 1.
-    %% Wait for the output vj for each RBCj such that j∈C. Finally output ∪j∈Cvj.
-    %% Note that this means if a BBA has returned 0, we don't need to wait for the corresponding RBC.
-    case lists:all(fun({E, RBC}) -> get_bba_result(Data, E) == false orelse rbc_completed(RBC) end, maps:to_list(Data#acs_data.rbc)) andalso
-         lists:all(fun(BBA) -> bba_completed(BBA) end, maps:values(Data#acs_data.bba)) of
-        true ->
-            ResultVector = [ {E, get_rbc_result(Data, E)} || E <- lists:seq(0, N-1), get_bba_result(Data, E) ],
-            {Data, {result_and_send, ResultVector, {send, ToSend}}};
+    %% once all instances of BA have completed, let C ⊂ [1..N] be the indexes of each BA that delivered 1.
+    %% Wait for the output vj for each RBCj such that j ∈ C. Finally output ∪j ∈ Cvj.
+    %% XXX: What? Note that this means if a BBA has returned 0, we don't need to wait for the corresponding RBC. What?
```

Comment:
the BBA returning 0 means that a majority have agreed the RBC has failed and there's no point in waiting for it.

Reply:
I see, yes. I think we wouldn't be doing that in the acc below anyway.